File: typescript/src/github-integrator.ts
/**
 * GitHub Integrator for creating pull requests with fix proposals
 */

import { Octokit } from '@octokit/rest';
import { Diagnosis } from './ai-diagnostician';
import { FixProposal } from './code-generator';

/**
 * GitHub configuration
 */
export interface GitHubConfig {
  token: string;
  owner: string;
  repo: string;
  baseBranch?: string;
  branchPrefix?: string;
  labels?: string[];
  assignees?: string[];
}

/**
 * Pull request information
 */
export interface PullRequest {
  number: number;
  url: string;
  branch: string;
}

/**
 * GitHub Integrator class for creating pull requests
 */
export class GitHubIntegrator {
  private octokit: Octokit;
  private config: GitHubConfig;
  private readonly DEFAULT_BASE_BRANCH = 'main';
  private readonly DEFAULT_BRANCH_PREFIX = 'vigilai-fix';

  constructor(config: GitHubConfig) {
    this.config = {
      ...config,
      baseBranch: config.baseBranch || this.DEFAULT_BASE_BRANCH,
      branchPrefix: config.branchPrefix || this.DEFAULT_BRANCH_PREFIX,
      labels: config.labels || [],
      assignees: config.assignees || [],
    };

    this.octokit = new Octokit({
      auth: this.config.token,
    });
  }

  /**
   * Create a pull request for a fix proposal
   */
  async createFixPR(
    proposal: FixProposal,
    diagnosis: Diagnosis
  ): Promise<PullRequest> {
    try {
      // Create a new branch
      const branchName = await this.createBranch(proposal.incidentId);

      // Commit changes to the branch
      await this.commitChanges(branchName, proposal);

      // Create pull request
      const pr = await this.createPullRequest(branchName, proposal, diagnosis);

      return pr;

    } catch (error) {
      console.error('Failed to create GitHub PR:', error instanceof Error ? error.message : 'Unknown error');
      throw error;
    }
  }

  /**
   * Create a new branch for the fix
   */
  private async createBranch(incidentId: string): Promise<string> {
    try {
      // Get the base branch reference
      const { data: baseRef } = await this.octokit.git.getRef({
        owner: this.config.owner,
        repo: this.config.repo,
        ref: `heads/${this.config.baseBranch}`,
      });

      // Create branch name with timestamp to ensure uniqueness
      const timestamp = Date.now();
      const branchName = `${this.config.branchPrefix}/${incidentId}-${timestamp}`;

      // Create new branch
      await this.octokit.git.createRef({
        owner: this.config.owner,
        repo: this.config.repo,
        ref: `refs/heads/${branchName}`,
        sha: baseRef.object.sha,
      });

      return branchName;

    } catch (error) {
      if (error instanceof Error && 'status' in error && (error as any).status === 422) {
        // Branch already exists, append additional timestamp
        const newTimestamp = Date.now();
        const branchName = `${this.config.branchPrefix}/${incidentId}-${newTimestamp}`;
        
        const { data: baseRef } = await this.octokit.git.getRef({
          owner: this.config.owner,
          repo: this.config.repo,
          ref: `heads/${this.config.baseBranch}`,
        });

        await this.octokit.git.createRef({
          owner: this.config.owner,
          repo: this.config.repo,
          ref: `refs/heads/${branchName}`,
          sha: baseRef.object.sha,
        });

        return branchName;
      }
      throw error;
    }
  }

  /**
   * Commit file changes to the branch
   */
  private async commitChanges(
    branchName: string,
    proposal: FixProposal
  ): Promise<void> {
    try {
      // Get the latest commit SHA for the branch
      const { data: refData } = await this.octokit.git.getRef({
        owner: this.config.owner,
        repo: this.config.repo,
        ref: `heads/${branchName}`,
      });

      const latestCommitSha = refData.object.sha;

      // Get the tree for the latest commit
      const { data: commitData } = await this.octokit.git.getCommit({
        owner: this.config.owner,
        repo: this.config.repo,
        commit_sha: latestCommitSha,
      });

      const baseTreeSha = commitData.tree.sha;

      // Create blobs for each file change
      const tree = await Promise.all(
        proposal.changes.map(async (change) => {
          const { data: blob } = await this.octokit.git.createBlob({
            owner: this.config.owner,
            repo: this.config.repo,
            content: Buffer.from(change.newContent).toString('base64'),
            encoding: 'base64',
          });

          return {
            path: change.path,
            mode: '100644' as const,
            type: 'blob' as const,
            sha: blob.sha,
          };
        })
      );

      // Create a new tree
      const { data: newTree } = await this.octokit.git.createTree({
        owner: this.config.owner,
        repo: this.config.repo,
        base_tree: baseTreeSha,
        tree,
      });

      // Create a new commit
      const { data: newCommit } = await this.octokit.git.createCommit({
        owner: this.config.owner,
        repo: this.config.repo,
        message: `Fix: ${proposal.description}`,
        tree: newTree.sha,
        parents: [latestCommitSha],
      });

      // Update the branch reference
      await this.octokit.git.updateRef({
        owner: this.config.owner,
        repo: this.config.repo,
        ref: `heads/${branchName}`,
        sha: newCommit.sha,
      });

    } catch (error) {
      console.error('Failed to commit changes:', error);
      throw error;
    }
  }

  /**
   * Create a pull request
   */
  private async createPullRequest(
    branchName: string,
    proposal: FixProposal,
    diagnosis: Diagnosis
  ): Promise<PullRequest> {
    try {
      // Build PR title and body
      const title = this.buildPRTitle(proposal, diagnosis);
      const body = this.buildPRBody(proposal, diagnosis);

      // Create the pull request
      const { data: pr } = await this.octokit.pulls.create({
        owner: this.config.owner,
        repo: this.config.repo,
        title,
        body,
        head: branchName,
        base: this.config.baseBranch!,
      });

      // Add labels if configured
      if (this.config.labels && this.config.labels.length > 0) {
        await this.octokit.issues.addLabels({
          owner: this.config.owner,
          repo: this.config.repo,
          issue_number: pr.number,
          labels: this.config.labels,
        });
      }

      // Add assignees if configured
      if (this.config.assignees && this.config.assignees.length > 0) {
        await this.octokit.issues.addAssignees({
          owner: this.config.owner,
          repo: this.config.repo,
          issue_number: pr.number,
          assignees: this.config.assignees,
        });
      }

      return {
        number: pr.number,
        url: pr.html_url,
        branch: branchName,
      };

    } catch (error) {
      console.error('Failed to create pull request:', error);
      throw error;
    }
  }

  /**
   * Build PR title
   */
  private buildPRTitle(proposal: FixProposal, diagnosis: Diagnosis): string {
    const category = diagnosis.category.replace('_', ' ');
    return `Fix: ${category} - ${diagnosis.rootCause.substring(0, 80)}`;
  }

  /**
   * Build PR body
   */
  private buildPRBody(proposal: FixProposal, diagnosis: Diagnosis): string {
    const lines: string[] = [
      `# Fix: ${diagnosis.category.replace('_', ' ')} Issue`,
      '',
      '## Incident Details',
      `- **ID**: ${proposal.incidentId}`,
      `- **Detected**: ${new Date(diagnosis.timestamp).toISOString()}`,
      `- **Category**: ${diagnosis.category}`,
      `- **Confidence**: ${(diagnosis.confidence * 100).toFixed(0)}%`,
      '',
      '## Diagnosis',
      diagnosis.explanation,
      '',
      '## Root Cause',
      diagnosis.rootCause,
      '',
    ];

    // Add code location if available
    if (diagnosis.codeLocation) {
      lines.push('## Code Location');
      lines.push(`- **File**: ${diagnosis.codeLocation.file}`);
      if (diagnosis.codeLocation.line) {
        lines.push(`- **Line**: ${diagnosis.codeLocation.line}`);
      }
      if (diagnosis.codeLocation.function) {
        lines.push(`- **Function**: ${diagnosis.codeLocation.function}`);
      }
      lines.push('');
    }

    // Add changes
    lines.push('## Changes');
    if (proposal.changes.length > 0) {
      proposal.changes.forEach(change => {
        lines.push(`- Modified: ${change.path}`);
      });
    } else {
      lines.push('No file changes generated.');
    }
    lines.push('');

    // Add suggested fix if available
    if (diagnosis.suggestedFix) {
      lines.push('## Suggested Fix');
      lines.push(diagnosis.suggestedFix);
      lines.push('');
    }

    // Add test cases if available
    if (proposal.testCases && proposal.testCases.length > 0) {
      lines.push('## Testing');
      proposal.testCases.forEach(testCase => {
        lines.push(`- ${testCase}`);
      });
      lines.push('');
    }

    lines.push('---');
    lines.push('*Generated automatically by VigilAI SDK*');

    return lines.join('\n');
  }
}

File: python/src/vigilai/github_integrator.py
"""GitHub Integrator for creating pull requests with fix proposals"""

from typing import Dict, List, Optional
from dataclasses import dataclass
from datetime import datetime
from github import Github, GithubException
from .ai_diagnostician import Diagnosis
from .code_generator import FixProposal


@dataclass
class GitHubConfig:
    """GitHub configuration"""
    token: str
    owner: str
    repo: str
    base_branch: Optional[str] = None
    branch_prefix: Optional[str] = None
    labels: Optional[List[str]] = None
    assignees: Optional[List[str]] = None


@dataclass
class PullRequest:
    """Pull request information"""
    number: int
    url: str
    branch: str


class GitHubIntegrator:
    """GitHub Integrator class for creating pull requests"""
    
    DEFAULT_BASE_BRANCH = 'main'
    DEFAULT_BRANCH_PREFIX = 'vigilai-fix'
    
    def __init__(self, config: GitHubConfig):
        """Initialize GitHub integrator with configuration"""
        self.config = config
        
        # Set defaults
        if not self.config.base_branch:
            self.config.base_branch = self.DEFAULT_BASE_BRANCH
        if not self.config.branch_prefix:
            self.config.branch_prefix = self.DEFAULT_BRANCH_PREFIX
        if not self.config.labels:
            self.config.labels = []
        if not self.config.assignees:
            self.config.assignees = []
        
        # Initialize GitHub client
        self.github = Github(self.config.token)
        self.repository = self.github.get_repo(f"{self.config.owner}/{self.config.repo}")
    
    def create_fix_pr(
        self,
        proposal: FixProposal,
        diagnosis: Diagnosis
    ) -> PullRequest:
        """Create a pull request for a fix proposal"""
        try:
            # Create a new branch
            branch_name = self._create_branch(proposal.incident_id)
            
            # Commit changes to the branch
            self._commit_changes(branch_name, proposal)
            
            # Create pull request
            pr = self._create_pull_request(branch_name, proposal, diagnosis)
            
            return pr
            
        except Exception as error:
            print(f"Failed to create GitHub PR: {str(error)}")
            raise
    
    def _create_branch(self, incident_id: str) -> str:
        """Create a new branch for the fix"""
        try:
            # Get the base branch reference
            base_ref = self.repository.get_git_ref(f"heads/{self.config.base_branch}")
            base_sha = base_ref.object.sha
            
            # Create branch name with timestamp to ensure uniqueness
            timestamp = int(datetime.now().timestamp() * 1000)
            branch_name = f"{self.config.branch_prefix}/{incident_id}-{timestamp}"
            
            # Create new branch
            self.repository.create_git_ref(
                ref=f"refs/heads/{branch_name}",
                sha=base_sha
            )
            
            return branch_name
            
        except GithubException as error:
            if error.status == 422:
                # Branch already exists, append additional timestamp
                new_timestamp = int(datetime.now().timestamp() * 1000)
                branch_name = f"{self.config.branch_prefix}/{incident_id}-{new_timestamp}"
                
                base_ref = self.repository.get_git_ref(f"heads/{self.config.base_branch}")
                base_sha = base_ref.object.sha
                
                self.repository.create_git_ref(
                    ref=f"refs/heads/{branch_name}",
                    sha=base_sha
                )
                
                return branch_name
            raise
    
    def _commit_changes(
        self,
        branch_name: str,
        proposal: FixProposal
    ) -> None:
        """Commit file changes to the branch"""
        try:
            # Get the latest commit SHA for the branch
            ref = self.repository.get_git_ref(f"heads/{branch_name}")
            latest_commit_sha = ref.object.sha
            
            # Get the tree for the latest commit
            commit = self.repository.get_git_commit(latest_commit_sha)
            base_tree_sha = commit.tree.sha
            
            # Create blobs for each file change
            tree_elements = []
            for change in proposal.changes:
                # Create blob
                blob = self.repository.create_git_blob(
                    content=change.new_content,
                    encoding='utf-8'
                )
                
                tree_elements.append({
                    'path': change.path,
                    'mode': '100644',
                    'type': 'blob',
                    'sha': blob.sha
                })
            
            # Create a new tree
            new_tree = self.repository.create_git_tree(
                tree=tree_elements,
                base_tree=self.repository.get_git_tree(base_tree_sha)
            )
            
            # Create a new commit
            new_commit = self.repository.create_git_commit(
                message=f"Fix: {proposal.description}",
                tree=new_tree,
                parents=[self.repository.get_git_commit(latest_commit_sha)]
            )
            
            # Update the branch reference
            ref.edit(sha=new_commit.sha)
            
        except Exception as error:
            print(f"Failed to commit changes: {str(error)}")
            raise
    
    def _create_pull_request(
        self,
        branch_name: str,
        proposal: FixProposal,
        diagnosis: Diagnosis
    ) -> PullRequest:
        """Create a pull request"""
        try:
            # Build PR title and body
            title = self._build_pr_title(proposal, diagnosis)
            body = self._build_pr_body(proposal, diagnosis)
            
            # Create the pull request
            pr = self.repository.create_pull(
                title=title,
                body=body,
                head=branch_name,
                base=self.config.base_branch
            )
            
            # Add labels if configured
            if self.config.labels and len(self.config.labels) > 0:
                pr.add_to_labels(*self.config.labels)
            
            # Add assignees if configured
            if self.config.assignees and len(self.config.assignees) > 0:
                pr.add_to_assignees(*self.config.assignees)
            
            return PullRequest(
                number=pr.number,
                url=pr.html_url,
                branch=branch_name
            )
            
        except Exception as error:
            print(f"Failed to create pull request: {str(error)}")
            raise
    
    def _build_pr_title(self, proposal: FixProposal, diagnosis: Diagnosis) -> str:
        """Build PR title"""
        category = diagnosis.category.replace('_', ' ')
        return f"Fix: {category} - {diagnosis.root_cause[:80]}"
    
    def _build_pr_body(self, proposal: FixProposal, diagnosis: Diagnosis) -> str:
        """Build PR body"""
        lines = [
            f"# Fix: {diagnosis.category.replace('_', ' ')} Issue",
            "",
            "## Incident Details",
            f"- **ID**: {proposal.incident_id}",
            f"- **Detected**: {datetime.fromtimestamp(diagnosis.timestamp / 1000).isoformat()}",
            f"- **Category**: {diagnosis.category}",
            f"- **Confidence**: {int(diagnosis.confidence * 100)}%",
            "",
            "## Diagnosis",
            diagnosis.explanation,
            "",
            "## Root Cause",
            diagnosis.root_cause,
            "",
        ]
        
        # Add code location if available
        if diagnosis.code_location:
            lines.append("## Code Location")
            lines.append(f"- **File**: {diagnosis.code_location.file}")
            if diagnosis.code_location.line:
                lines.append(f"- **Line**: {diagnosis.code_location.line}")
            if diagnosis.code_location.function:
                lines.append(f"- **Function**: {diagnosis.code_location.function}")
            lines.append("")
        
        # Add changes
        lines.append("## Changes")
        if len(proposal.changes) > 0:
            for change in proposal.changes:
                lines.append(f"- Modified: {change.path}")
        else:
            lines.append("No file changes generated.")
        lines.append("")
        
        # Add suggested fix if available
        if diagnosis.suggested_fix:
            lines.append("## Suggested Fix")
            lines.append(diagnosis.suggested_fix)
            lines.append("")
        
        # Add test cases if available
        if proposal.test_cases and len(proposal.test_cases) > 0:
            lines.append("## Testing")
            for test_case in proposal.test_cases:
                lines.append(f"- {test_case}")
            lines.append("")
        
        lines.append("---")
        lines.append("*Generated automatically by VigilAI SDK*")
        
        return "\n".join(lines)
